soc/arm/nuvoton_npcx/npcx7/power.c

```diff
@@ -0,0 +1,157 @@
+ *
+ * Based on the follwoing criteria:
+ *
+ * - A delay of 'Instant' wake-up from 'Deep Sleep' is 20 us.
+ * - A delay of 'Standard' wake-up from 'Deep Sleep' is 3.43 ms.
+ * - Max residency time in Deep Sleep for 'Instant' wake-up is 200 ms
+ * - Min Residency time in Deep Sleep for 'Instant' wake-up is 61 us
+ * - The unit to determine power state residency policy is tick.
+ *
+ * this driver implements one power state, PM_STATE_SUSPEND_TO_IDLE, with
+ * two sub-states for power management system.
+ * Sub-state 0 - "Deep Sleep" mode with “Instant” wake-up if residency time
+ *               is greater or equal to 1 ms
+ * Sub-state 1 - "Deep Sleep" mode with "Standard" wake-up if residency time
+ *               is greater or equal to 201 ms
+ *
+ * INCLUDE FILES: soc_clock.h
+ */
+
+#include <zephyr.h>
+#include <power/power.h>
+#include <soc.h>
+
+#include <logging/log.h>
+LOG_MODULE_DECLARE(soc, CONFIG_SOC_LOG_LEVEL);
+
+/* The steps that npcx ec enters sleep/deep mode and leaves it. */
+#define NPCX_ENTER_SYSTEM_SLEEP() ({                                           \
+	__asm__ volatile (                                                     \
+	"push {r0-r5}\n"     /* Save the registers used for delay */           \
+	"wfi\n"              /* Enter sleep mode after receiving wfi */        \
+	"ldm %0, {r0-r5}\n"  /* Add a delay before instructions fetching */    \
+	"pop {r0-r5}\n"      /* Restore the registers used for delay */        \
+	"isb\n"              /* Flush the cpu pipelines */                     \
+	:: "r" (CONFIG_SRAM_BASE_ADDRESS)); /* A valid addr used for delay */  \
+	})
+
+#if defined(CONFIG_SOC_POWER_MANAGEMENT_TRACE)
+uint32_t cnt_sleep0;
+uint32_t cnt_sleep1;
+#endif
+
+static void npcx_power_enter_system_sleep(bool is_deep, bool is_instant)
+{
+	/* Disable interrupts */
+	__disable_irq();
+
+	/*
+	 * Disable priority mask temporarily to make sure that wake-up events
+	 * are visible to the WFI instruction.
+	 */
+	__set_BASEPRI(0);
+
+	/* Configure sleep/deep sleep settings in clock control module. */
+	npcx_clock_control_turn_on_system_sleep(is_deep, is_instant);
+
+	/*
+	 * Capture the reading of low-freq timer for compensation before ec
+	 * enters system sleep mode.
+	 */
+	npcx_clock_capture_low_freq_timer();
+
+	/* Enter system sleep mode */
+	NPCX_ENTER_SYSTEM_SLEEP();
+
+	/*
+	 * Compensate system timer by the elasped time of low-freq timer during
+	 * system sleep mode.
+	 */
+	npcx_clock_compensate_system_timer();
+
+	/* Turn off system sleep mode. */
+	npcx_clock_control_turn_off_system_sleep();
+}
+
+/* Invoke when enter "Suspend/Low Power" mode. */
+void pm_power_state_set(struct pm_state_info info)
+{
+	if (info.state != PM_STATE_SUSPEND_TO_IDLE) {
+		LOG_DBG("Unsupported power state %u", info.state);
+	} else {
+		switch (info.substate_id) {
+		case 0:	/* Sub-state 0: Deep sleep with instant wake-up */
+			npcx_power_enter_system_sleep(1, 1);
```

Comment:
If you change the parameters to npcx_power_enter_system_sleep() to enums, then this code becomes a lot more readable.  For example:
    npcx_power_enter_system_sleep(DEEP_SLEEP, INSTANT_WAKE_UP);
    npcx_power_enter_system_sleep(DEEP_SLEEP, STANDARD_WAKE_UP);

Reply:
Done.